src/service/url-replacements-impl.js

```diff
@@ -74,6 +74,22 @@ export class GlobalVariableSource extends VariableSource {
     /** @const {!./ampdoc-impl.AmpDoc} */
     this.ampdoc = ampdoc;
 
+    // A meta[name="amp-action-whitelist"] tag, if present, contains,
+    // in its content attribute, a whitelist of actions on the special AMP target.
+    if (this.ampVariableSubstitutionWhitelist_ === undefined
+      && this.ampdoc.getRootNode() && this.ampdoc.getRootNode().head) {
+      const meta =
+        this.ampdoc.getRootNode().head
+            .querySelector('meta[name="amp-variable-substitution-whitelist"]');
+
+      // Cache the whitelist of allowed AMP actions (if provided).
+      if (meta) {
+        this.ampVariableSubstitutionWhitelist_ =
+          meta.getAttribute('content').split(',')
+              .map(action => action.trim());
```

Comment:
I see that this code is repeated from the previous PR. Consider extracting this to a utility class.